tensorflow/java/src/main/java/org/tensorflow/Shape.java

```diff
@@ -74,7 +74,41 @@ public int numDimensions() {
    * @return The size of the requested dimension or -1 if it is unknown.
    */
   public long size(int i) {
-    return shape[i];
+    return shape == null ? -1 : shape[i];
+  }
+
+  /**
+   * The total number of elements found in a tensor of this shape.
+   * 
+   * <p>If the size of some dimensions is unknown, the total number of elements cannot be calculated and -1 is returned.
+   * 
+   * @return the number of elements or -1 if size of some dimension are unknown
+   */
+  public int numElements() {
+    if (shape == null) {
+       return -1;
+    }
+    int total = 1;
+    for (int i = 0; i < shape.length; ++i) {
+      // TODO (karllessard): There might be a lossy conversion here from 'long' sizes to 'int' total, but this issue
+      // seems ubiquitous in the current Java client implementation. It should be adressed all at once.
+      int size = (int) size(i);
+      if (size < 0) {
+        return -1;
+      }
+      total *= size;
+    }
+    return total;
+  }
+
+  /**
+   * Returns the shape as an array.
+   * 
+   * <p>Each element represent the size of the dimension at the given index. For example,
+   * {@code shape.asArray()[2]} is equal to the size of the third dimension of this shape.
+   */
+  public long[] asArray() {
```

Comment:
I'd like to figure out a way to avoid making this public.
I was intentionally avoiding doing this (because this enables a public API to mutate the shape which can be problematic).

Reply:
That one bugged me too, but I had to because the Tensor interface only accepts as arrays.

IMHO, I right way to address this is to modify the `Tensor` interface as well to accept their shapes as `Shape` objects instead. Actually, I did it and then revert it to keep my PR small :)

Since we do not guarantee API stability for the Java client yet, can we just change the signature of those methods in `Tensor` or should we add a new variant for each of them?

Reply:
`Tensor` accepts `long[]` and not `Shape` objects in the spirit of making it clear that `Tensor`s must be created with fully known shapes (and `Shape` objects can also represent partially known shapes). For the same reason, `Tensor.shape` returns a `long[]` and not a `Shape`.

Thus, I'm not a big fan of changing the signature of the `Tensor` factory methods to accept partially known shapes.